quantization/src/encoded_vectors_pq.rs

```diff
@@ -0,0 +1,575 @@
+        for _ in 0..len / 4 {
+            let buffer = [
+                *lut.add(*centroids as usize),
+                *lut.add(centroids_count + *centroids.add(1) as usize),
+                *lut.add(2 * centroids_count + *centroids.add(2) as usize),
+                *lut.add(3 * centroids_count + *centroids.add(3) as usize),
+            ];
+            let c = _mm_loadu_ps(buffer.as_ptr());
+            sum128 = _mm_add_ps(sum128, c);
+
+            centroids = centroids.add(4);
+            lut = lut.add(4 * centroids_count);
+        }
+        let sum64: __m128 = _mm_add_ps(sum128, _mm_movehl_ps(sum128, sum128));
+        let sum32: __m128 = _mm_add_ss(sum64, _mm_shuffle_ps(sum64, sum64, 0x55));
+        let mut sum = _mm_cvtss_f32(sum32);
+
+        for _ in 0..len % 4 {
+            sum += *lut.add(*centroids as usize);
+            centroids = centroids.add(1);
+            lut = lut.add(centroids_count);
+        }
+        sum
+    }
+
+    #[cfg(all(target_arch = "aarch64", target_feature = "neon"))]
+    unsafe fn score_point_neon(&self, query: &EncodedQueryPQ, i: u32) -> f32 {
+        let centroids = self
+            .encoded_vectors
+            .get_vector_data(i as usize, self.metadata.vector_division.len());
+        let len = centroids.len();
+        let centroids_count = self.metadata.centroids.len();
+
+        let mut centroids = centroids.as_ptr();
+        let mut lut = query.lut.as_ptr();
+        let mut sum128 = vdupq_n_f32(0.);
+        for _ in 0..len / 4 {
+            let buffer = [
+                *lut.add(*centroids as usize),
+                *lut.add(centroids_count + *centroids.add(1) as usize),
+                *lut.add(2 * centroids_count + *centroids.add(2) as usize),
+                *lut.add(3 * centroids_count + *centroids.add(3) as usize),
+            ];
+            let c = vld1q_f32(buffer.as_ptr());
+            sum128 = vaddq_f32(sum128, c);
+
+            centroids = centroids.add(4);
+            lut = lut.add(4 * centroids_count);
+        }
+        let mut sum = vaddvq_f32(sum128);
+
+        for _ in 0..len % 4 {
+            sum += *lut.add(*centroids as usize);
+            centroids = centroids.add(1);
+            lut = lut.add(centroids_count);
+        }
+        sum
+    }
+
+    fn score_point_simple(&self, query: &EncodedQueryPQ, i: u32) -> f32 {
+        unsafe {
+            let centroids = self
+                .encoded_vectors
+                .get_vector_data(i as usize, self.metadata.vector_division.len());
+            let len = centroids.len();
+            let centroids_count = self.metadata.centroids.len();
+
+            let mut centroids = centroids.as_ptr();
+            let mut lut = query.lut.as_ptr();
+
+            let mut sum = 0.0;
+            for _ in 0..len {
+                sum += *lut.add(*centroids as usize);
+                centroids = centroids.add(1);
+                lut = lut.add(centroids_count);
+            }
+            sum
```

Comment:
Limit `unsafe` block to unsafe calls, and us a `sum` iterator:

```suggestion
        let centroids = self
            .encoded_vectors
            .get_vector_data(i as usize, self.metadata.vector_division.len());
        let len = centroids.len();
        let centroids_count = self.metadata.centroids.len();

        let mut centroids = centroids.as_ptr();
        let mut lut = query.lut.as_ptr();
        
        (0..len)
            .map(|_| unsafe {
                let value = *lut.add(*centroids as usize);
                centroids = centroids.add(1);
                lut = lut.add(centroids_count);
                value
            })
            .sum()
```

Reply:
agreed